### DISCUSSION TITLE
Meshwarp images

### DISCUSSION THREAD

**Discussion by DOEHOONLEE** [ImageMagick/ImageMagick#4130] [Help]
Hi! I was wondering if I can distort/warp the image according to the warp points that I got on Photoshop.

I did learn some ways to warp an image with
```
-compose Displace

-distort Perspective

-shear

-wave
```
but unless I have the exact values, it is extremely hard to get the exactly warped image I want. So I am trying to warp with the data I have. Is this achievable?

Thanks! Here's the reference image with the points that I have.

<img width="425" alt="Screen Shot 2021-08-30 at 5 33 18 PM" src="https://user-images.githubusercontent.com/20305442/131310780-e5706d6d-70a5-4e52-a128-39de552dca76.png">

**Comment by snibgo** [comment]
"-distort Shepards" may do what you want. See https://imagemagick.org/script/command-line-options.php#distort

It looks as if you maybe you want to distort a rectangle to a coffee mug, which is a cylinder. If so, that is a mathematical transformation. See [Coffee mug](http://im.snibgo.com/coffmug.htm).

Another possibility is to distort a rectangle so the top and bottom edges become the required curves. However, this will not do the horizontal distortion to "shrink" the left and right edges. See [Straightening two lines](http://im.snibgo.com/str2lines.htm), but you would do the inverse operation.

Another possibility is [Curving in 3D](http://im.snibgo.com/curve3d.htm), which is like curving a piece of paper in 3D and projecting the image to a plane surface.

EDIT: If you have a single image that you want to tweak, Gimp or Photoshop may be better options.

You might post whatever inputs you have, for more specific advice.

**Reply by DOEHOONLEE** [reply]
oh~ yes! I forgot to mention. I have tried the "-distort Shepards" one, but I could not get the exact shape that I want.

<img width="632" alt="Screen Shot 2021-08-31 at 7 59 17 AM" src="https://user-images.githubusercontent.com/20305442/131416412-01b40f9e-c36d-41c7-bbdc-d5c0c758a406.png">

I had to keep adjusting the values so that the image fits in the mug (roughly by my eyes). Assuming that I have drawn the mug, I will have all the values/dimensions for it. Then with those input values, perhaps I could output the exactly warped image that I want?

Would that be possible? Would this method work on mapping blankets too?

Also, thanks for introducing the [Coffee Mug](http://im.snibgo.com/coffmug.htm). It is awesome! Though, um, can I run those on servers? is there node.js version of it?

**Comment by fmw42** [comment]
You can try a second order polynomial warp, if you have at least 12 control points.  See also https://legacy.imagemagick.org/Usage/mapping/#displace_cylinder. If on a Unix-like system (Linux, Mac OSX, Windows 10 Unix), you can try my cylinderize script, though you need some trial an error to get it to fit well.

**Reply by DOEHOONLEE** [reply]
Wow! That is awesome!

```
convert rose: -background black -gravity south -splice 0x8 \
          \( +clone -sparse-color barycentric '0,0 black 69,0 white' \) \
          \( -clone 1 -function arcsin 0.5 \) \
          \( -clone 1 -level 25%,75% \
                 -function polynomial -4,4,0 -gamma 2 \
                 +level 50%,0 \) \
          -delete 1 \
          -virtual-pixel black  -define compose:args=17x7 \
          -compose Displace  -composite   rose_cylinder.png
```

Okay. So I assume that `-gravity south` pulls down the image downwards so that it is arch shaped. Why are there `+` and `-` signs for the same command?
Are there explanations for these commands?

I want to apply the exact values(warped/transformed) of the mug so that I won't have to guess the inputs when runningi `convert` command. Do you, by any chance, know a way to do so?

I do have the points and I can probably calculate degrees of the arch shape, but how can I convert them into ... these... `-gamma` or `-level` and stuff..?

What are those commands for? I read the description right below, but they are not that straightforward.

Thanks!

**Reply by fmw42** [reply]
The arcsin does the cylinder wrap and polynomial does the effective tilt. So there is no way to directly map your points to that kind of transformation. -gravity south directs the splice where to add rows of black pixels. +gravity resets the gravity setting to its default value.  -gamma 2 is like a square root. +level compresses the gray levels. The work is done by the displacement mapping form -compose displace -composite according to the mapping from the arcsin and polynomial.

**Reply by DOEHOONLEE** [reply]
omg.. I am not too sure what you meant by cylinder wrap and polynomial :(
but I get that they are the standards where the `-compose Displace` use to map the image around the mug. Do you know what those values represent? on the image? Perhaps, I could come up with formulas.

**Comment by fmw42** [comment]
The arcsin is the mathematical expression used to warp horizontally in a cylinder. The -distort polynomial is the math to warp vertically so that the cylinder looks tilted.  What values are you asking about?

**Reply by DOEHOONLEE** [reply]
Can you explain what `-clone` is and why one starts with `-` and another starting with `+`? Thanks!!

**Reply by DOEHOONLEE** [reply]
Hmm... So I found this `Many of the options have both a '-' and a '+' style. The latter is generally used to turn off the setting, or reset it to its normal default state. This allow you remove the effect of a setting quickly and simply` but..
```
( +clone -sparse-color barycentric '0,0 black 69,0 white' \) \
          \( -clone 1 -function arcsin 0.5 \) \
          \( -clone 1 -level 25%,75% \
```
in here, why would the first one have `+` with options? hmm..

**Reply by DOEHOONLEE** [reply]
Never mind, I found them!!
![Screen Shot 2021-08-31 at 11 50 16 AM](https://user-images.githubusercontent.com/20305442/131433766-93ae99c7-7aa7-43cf-b991-0a53af55fca9.png)

**Comment by fmw42** [comment]
-clone 1 copies the second image (numbering from 0) in the command line. +clone copies the previous image and has no numbering associated with it.  See https://legacy.imagemagick.org/Usage/basics/#clone
